bigbluebutton-html5/imports/api/2.0/annotations/server/modifiers/addAnnotation.js

```diff
@@ -18,46 +42,315 @@ export default function addAnnotation(meetingId, whiteboardId, userId, annotatio
+          };
+
+        // if lastChunkLength > PENCIL_CHUNK_SIZE then we need to create another chunk
+        } else if (lastChunkLength >= PENCIL_CHUNK_SIZE) {
+          baseModifier = {
+            $set: {
+              numberOfChunks: numberOfChunks + 1,
+              lastChunkLength: annotationInfo.points.length,
+              lastCoordinate: [
+                annotationInfo.points[annotationInfo.points.length - 2],
+                annotationInfo.points[annotationInfo.points.length - 1],
+              ],
+            },
+          };
+
+          const chunkId = `${id}--${numberOfChunks + 1}`;
+          chunkSelector = {
+            meetingId,
+            userId,
+            id: chunkId,
+          };
+
+          // pushing the last coordinate to the front of the current chunk's points
+          annotationInfo.points.unshift(Annotation.lastCoordinate[0], Annotation.lastCoordinate[1]);
+
+          chunkModifier = {
+            $set: {
+              whiteboardId,
+              meetingId,
+              userId,
+              id: chunkId,
+              status,
+              annotationType,
+              annotationInfo,
+              wbId,
+              position: Annotation.position,
+            },
+            $inc: { version: 1 },
+          };
+        }
+
+        // upserting the new subdocument
+        Annotations.upsert(chunkSelector, chunkModifier);
+        // base will be updated in the main AddAnnotation func
+        return { selector: baseSelector, modifier: baseModifier };
+      }
+
+      // **default flow**
+      // if we are here then it means that Annotation object is not in the db
+      // So creating everything similar to DRAW_START case
+      const _chunks = createPencilObjects();
 
-  switch (shapeType) {
+      // creating 'pencil_base' based on the info we received from createPencilObjects()
+      baseModifier = {
+        id,
+        userId,
+        meetingId,
+        position,
+        annotationType: 'pencil_base',
+        numberOfChunks: _chunks.length,
+        lastChunkLength: _chunks[_chunks.length - 1].length,
+        lastCoordinate: [
+          annotationInfo.points[annotationInfo.points.length - 2],
+          annotationInfo.points[annotationInfo.points.length - 1],
+        ],
+      };
+
+      // upserting all the chunks
+      for (let i = 0; i < _chunks.length; i += 1) {
+        Annotations.upsert(_chunks[i].selector, _chunks[i].modifier);
+      }
+
+      // base will be updated in the main AddAnnotation func
+      return { selector: baseSelector, modifier: baseModifier };
+    }
+    case DRAW_END: {
+      // If a user just finished drawing with the pencil
+      // Removing all the sub-documents and replacing the 'pencil_base'
+      if (Annotation && Annotation.annotationType === 'pencil_base') {
+        // delete everything and replace base
+        const chunkIds = [];
+        for (let i = 0; i <= Annotation.numberOfChunks; i += 1) {
```

Comment:
```Array(Annotation.numberOfChunks).fill(null).map((_,i) => `${Annotation.id}--${i}`)```

Reply:
I think readability will suffer here 😕 

You are suggesting to:
create an array ->
fill it with `null`s ->
map each value to the string ->
assign the result to the variable

It's a bit confusing, isn't it? I find a standard for loop much easier to read.

Reply:
We agree to disagree then.

To me its way easier to read something that I can break in smaller steps:
- create an array
- fill it with nulls
- map each value to an string template

Maybe its because I'm more into functional programming than Imperative programming.

Reply:
Guys, these seem fairly equal alternatives. I am alright going forward with the existing (`for`) way

Reply:
@oswaldoacauan just a follow up on the subject.
Created a test case: https://jsperf.com/filling-an-array-with-values-imperative-vs-functional
The results in Firefox (57) didn't surprise me, my approach is 30-40% faster.
I anticipated these results, since there is no that unnecessary step, filling an array with `null`s.

But then I ran the same test in Chrome and was quite surprised with the result. It looks like Chrome has some issues with basic array (or string?) operations.

Reply:
P.S. Added a third test case to the link above ^^ - "direct assignment".
I think I need to make a change here now.

Reply:
@OZhurbenko do we care more about readability or minimal performance gains?
Because if thats the case, an while will be the best choice and we should not be testing agains browsers since this code will run in Node.
https://jsperf.com/filling-an-array-with-values-imperative-vs-functional-2

IMO we should care more about readability than those minimal performance gains. We should let transpilers and engines take care of that :)

Reply:
@oswaldoacauan  Our client is written in JavaScript. JS is a OO (+ imperative) language with some functional elements. Using pure functional approach in a language like that will always be significantly less readable. Unless we plan to refactor the code base in a functional way, which we don't. Or do we, @antobinary ?
Actually, even developers that use functional languages mostly agree it is less readable. But it makes sense sometimes, because there are significant advantages: testability and performance gains. 1) We can't really achieve any significant performance gains in JS. JavaScript is not Scala, it doesn't have things like `withFilter` and `view`. 2) Testability is not the case here either. If we want to make it more testable, we should use a pure function and recursion here.
Long story short, I'm all for the functional approach. But I think in that case, we can't just use it once or twice here and there. It must be a paradigm switch. Otherwise, we have the whole client object-oriented and then someone comes to this line and get stuck for 10 seconds just thinking about what's going on.
P.S. Want to throw Scala.js here, in case we plan to become fanatics.